src/Analysis/Ast/Impl/Types/PythonClassType.Generics.cs

```diff
@@ -0,0 +1,360 @@
+                    }
+                }
+            }
+
+            // For still undefined parameters try matching passed types in order
+            for (int i = 0, gtIndex = 0; i < args.Arguments.Count; i++) {
+                var arg = args.Arguments[i];
+
+                if (Equals(arg.Type)) {
+                    continue; // Typically 'self'.
+                }
+
+                if (arg.Value is IMember member) {
+                    var type = member.GetPythonType();
+                    if (!type.IsUnknown()) {
+                        var gtd = gtIndex < genericTypeParameters.Length ? genericTypeParameters[gtIndex] : null;
+                        if (gtd != null && genericTypeToSpecificType.TryGetValue(gtd, out var s) && s is IGenericTypeParameter) {
+                            genericTypeToSpecificType[gtd] = type;
+                        }
+                        gtIndex++;
+                    }
+                }
+            }
+            return genericTypeToSpecificType;
+        }
+
+        /// <summary>
+        /// Given an argument set with types, e.g int, float, str, look at the generic type parameters for the class
+        /// and create a new class type with those parameters filled in. Additionally, transmit the specific class
+        /// types to bases and recreate them with their specific types.
+        /// 
+        /// e.g
+        /// class A(Generic[T, U]): ...
+        /// class B(A[T, str]): ...
+        /// 
+        /// B[int] defines the type parameter T to be of type int and type parameter U to be type str. 
+        /// B[int] inherits from A[int, str] 
+        /// </summary>
+        public IPythonType CreateSpecificType(IArgumentSet args) {
+            var genericTypeParameters = GetTypeParameters();
+            var newBases = new List<IPythonType>();
+
+            // Get map of generic type parameter to specific type - fill in what type goes to what
+            // type parameter T -> int, U -> str, etc.
+            var genericTypeToSpecificType = GetSpecificTypes(args, genericTypeParameters, newBases);
+
+            PythonClassType classType = new PythonClassType(BaseName, new Location(DeclaringModule));
+            // Storing generic parameters allows methods returning generic types 
+            // to know what type parameter returns what specific type
+            StoreGenericParameters(classType, genericTypeParameters, genericTypeToSpecificType);
+
+            // Locking so only one thread access cache
+            lock (_genericParameterLock) {
+                // Store generic parameters first so name updates correctly, then check if class type has been cached
+                _specificTypeCache = _specificTypeCache ?? new Dictionary<string, PythonClassType>();
+                if (_specificTypeCache.TryGetValue(classType.Name, out var cachedType)) {
+                    return cachedType;
+                }
+                _specificTypeCache[classType.Name] = classType;
```

Comment:
So here is the problem with lock scope: thread 1 enters this and adds `classType` to the dictionary. Thread 2 enters and gets it from the dictionary. But class type hasn't been constructed fully yet - class bases and members are only set below, at line 234+. The lock has to fully protect type construction.

Reply:
I understand now, I will lock the initialization of the class then.

Reply:
Actually problem is below, when members/bases are set, not when the class is new'ed. Basically, all code from current `lock() {` to the end of the function has to be locked.

Reply:
Yes that is what I did, by initialization I meant setting the members and bases as well.